File: LeetCode/322.Coin Change.py
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        if amount == 0: return 0
        coins = list(reversed(sorted(coins)))
        dp, stack = {0:0}, [0]
        while len(stack) > 0:
            next = []
            for val in stack:
                for coin in coins:
                    next_val = coin + val
                    if next_val not in dp:
                        dp[next_val] = dp[val] + 1
                        if next_val == amount:
                            return dp[next_val]
                        if next_val < amount:
                            next.append(next_val)
            stack = next
        return -1

    def coinChange_TLE(self, coins, amount):
        dp = [-1] * (amount + 1)
        dp[0] = 0
        for i in range(1, amount + 1):
            dp[i] = amount + 1
            for j in range(len(coins)):
                val = i - coins[j]
                if val >= 0 and dp[val] != -1:
                    dp[i] = min(dp[val] + 1, dp[i])
            if dp[i] == amount + 1:
                dp[i] = -1
        return dp[amount]

File: 322. Coin Change.java
/**
You are given an integer array coins representing coins of different denominations and an integer amount representing a total amount of money.

Return the fewest number of coins that you need to make up that amount. If that amount of money cannot be made up by any combination of the coins, return -1.

You may assume that you have an infinite number of each kind of coin.

 

Example 1:

Input: coins = [1,2,5], amount = 11
Output: 3
Explanation: 11 = 5 + 5 + 1
Example 2:

Input: coins = [2], amount = 3
Output: -1
Example 3:

Input: coins = [1], amount = 0
Output: 0
 

Constraints:

1 <= coins.length <= 12
1 <= coins[i] <= 231 - 1
0 <= amount <= 104
*/

class Solution {
    public int coinChange(int[] coins, int amount) {
        if (amount == 0)
            return 0;
        Arrays.sort(coins);
        int[] minCounts = new int[amount + 1];
        for (int i = 1; i <= amount; i++) {
            int min = 100001;
            for (int j = coins.length - 1; j >= 0; j--) {
                int rest = i - coins[j];
                if (rest < 0)
                    continue;
                if (minCounts[rest] != -1) {
                    minCounts[i] = minCounts[rest] + 1;
                    if (minCounts[i] < min)
                        min = minCounts[i];
                }
            }
            if (min == 100001)
                minCounts[i] = -1;
            else 
                minCounts[i] = min;
        }
        
        return minCounts[amount] != 0 ? minCounts[amount]: -1;
    }
}
